### PR TITLE
Refactor CassandraRecordSink

### PR DESCRIPTION
**Pull request by arhimondr**
Build query with query builder provided by the DataStax driver.
Use PreparedStatement to perform inserts.
This will ensure proper column and table names escape, and improve
performance.

This is a pre-requisite for https://github.com/prestodb/presto/pull/7451

### CODE REVIEW

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java [hunk 1]
```diff
@@ -41,37 +49,36 @@
...
+        this.columnTypes = ImmutableList.copyOf(requireNonNull(columnTypes, "columnTypes is null"));
+        this.generateUUID = generateUUID;
 
-        String schemaName = handle.getSchemaName();
-        StringBuilder queryBuilder = new StringBuilder(String.format("INSERT INTO \"%s\".\"%s\"(", schemaName, handle.getTableName()));
-        queryBuilder.append("id");
-
-        for (String columnName : handle.getColumnNames()) {
-            queryBuilder.append(",").append(columnName);
+        Insert insert = insertInto(schemaName, tableName);
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: requireNonNull for schemaName and tableName

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java [hunk 2]
```diff
@@ -41,37 +49,36 @@
...
+        if (generateUUID) {
+            insert.value("id", bindMarker());
         }
-        queryBuilder.append(") VALUES (?");
-
-        for (int i = 0; i < handle.getColumnNames().size(); i++) {
-            queryBuilder.append(",?");
+        for (String columnName : columnNames) {
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: requireNonNull for columnNames

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java [hunk 3]
```diff
@@ -41,37 +49,36 @@
...
+            insert.value(columnName, bindMarker());
         }
-        queryBuilder.append(")");
+        this.insert = cassandraSession.executeWithSession(s -> s.prepare(insert));
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: When this is rebased on https://github.com/prestodb/presto/pull/7914, how will this work? You'll need to add another method to CassandraSession, right?

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java [hunk 4]
```diff
@@ -41,37 +49,36 @@
...
 
-        insertQuery = queryBuilder.toString();
-        values = new ArrayList<>();
-
-        columnTypes = handle.getColumnTypes();
+        values = new ArrayList<>(columnTypes.size() + 1);
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: This should only be +1 if you're adding the id

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java [hunk 5]
```diff
@@ -81,16 +88,18 @@ public void beginRecord()
...
 
     @Override
     public void finishRecord()
     {
         checkState(field != -1, "not in record");
-        checkState(field == fieldCount, "not all fields set");
+        checkState(field == columnTypes.size(), "not all fields set");
         field = -1;
-        cassandraSession.execute(insertQuery, values.toArray());
+        cassandraSession.executeWithSession(s -> s.execute(insert.bind(values.toArray())));
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: Ditto on the need to change CassandraSession when rebased on https://github.com/prestodb/presto/pull/7914

Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: Also, I don't see that Insert has a .bind() method in the java docs I'm looking at, but I trust that the code compiles; probably the more recent version of the driver includes that.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java [hunk 6]
```diff
@@ -108,18 +117,25 @@ public void appendBoolean(boolean value)
...
-        else if (INTEGER.equals(columnTypes.get(field))) {
+        else if (INTEGER.equals(columnType)) {
             append(((Number) value).intValue());
         }
-        else if (REAL.equals(columnTypes.get(field))) {
+        else if (REAL.equals(columnType)) {
             append(intBitsToFloat((int) value));
         }
-        else {
+        else if (TIMESTAMP.equals(columnType)) {
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: Separate out making TIMESTAMP and VARBINARY work properly, please

Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: Also, add tests for timestamp, making sure you're handling the value properly between Presto and Cassandra

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java [hunk 7]
```diff
@@ -131,7 +147,16 @@ public void appendDouble(double value)
     @Override
     public void appendString(byte[] value)
     {
-        append(new String(value, UTF_8));
+        Type columnType = columnTypes.get(field);
+        if (VARBINARY.equals(columnType)) {
+            append(ByteBuffer.wrap(value));
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: ditto on adding tests

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java (hunk 3 above)
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: I've exposed `prepare` method

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java (hunk 4 above)
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: As it is only a size hint it doesn't make a big difference. What matters is that we have pre-allocated array that fits all the columns.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java (hunk 5 above)
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: There is a `execute(Statement)` method exposed

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java (hunk 6 above)
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: We do have tests for that. The only difference is the way we bind values. Previously `execute(sql, arguments)` method that we used was more `smart`. That method was able to do more sophisticated types conversions.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java (hunk 7 above)
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: ditto

Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: `.bind()` method requires all the parameters to be a proper type of an actual column  type

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java (hunk 2 above)
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraRecordSink.java]: Each of the values in columnName need to be non-null, too

**Comment by cawallin** [comment]
Merged, thanks!

**Pull request merged**